patches/server/1006-PreSpawnerSearchEvent.patch

```diff
@@ -0,0 +1,66 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TreemanKing <67459602+TreemanKing@users.noreply.github.com>
+Date: Wed, 16 Aug 2023 11:53:44 +1000
+Subject: [PATCH] PreSpawnerSearchEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+index 633500aefd515df5dadda3802b94079f75a03fa0..44f7d7adb19eb42a76b97949ee1ba0754538d2ff 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -1,6 +1,8 @@
+ package net.minecraft.world.level;
+ 
+ import com.mojang.logging.LogUtils;
++
++import java.util.List;
+ import java.util.Optional;
+ import java.util.function.Function;
+ import javax.annotation.Nullable;
+@@ -16,12 +18,8 @@ import net.minecraft.util.RandomSource;
+ import net.minecraft.util.random.SimpleWeightedRandomList;
+ import net.minecraft.util.random.WeightedEntry;
+ import net.minecraft.world.Difficulty;
+-import net.minecraft.world.entity.Entity;
+-import net.minecraft.world.entity.EntityType;
+-import net.minecraft.world.entity.Mob;
+-import net.minecraft.world.entity.MobSpawnType;
+-import net.minecraft.world.entity.SpawnGroupData;
+-import net.minecraft.world.entity.SpawnPlacements;
++import net.minecraft.world.entity.*;
++import net.minecraft.world.entity.player.Player;
+ import net.minecraft.world.level.gameevent.GameEvent;
+ import net.minecraft.world.phys.AABB;
+ import org.slf4j.Logger;
+@@ -46,6 +44,7 @@ public abstract class BaseSpawner {
+     public int requiredPlayerRange = 16;
+     public int spawnRange = 4;
+     private int tickDelay = 0; // Paper
++    private int searchCooldownTicks = 0; // Paper
+ 
+     public BaseSpawner() {}
+ 
+@@ -86,6 +85,23 @@ public abstract class BaseSpawner {
+         tickDelay = world.paperConfig().tickRates.mobSpawner;
+         if (tickDelay == -1) { return; } // If disabled
+         // Paper end
++
++        // Paper start - PreSpawnerSearchEvent
++        if (--searchCooldownTicks > 0) return;
++
++        List<org.bukkit.entity.Player> activatingPlayers = (List)world.getNearbyPlayers(null,
++            (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double)
++                this.requiredPlayerRange, EntitySelector.PLAYER_AFFECTS_SPAWNING);
++
++        com.destroystokyo.paper.event.entity.PreSpawnerSearchEvent preSpawnerSearchEvent;
++        preSpawnerSearchEvent = new com.destroystokyo.paper.event.entity.PreSpawnerSearchEvent(
++            io.papermc.paper.util.MCUtil.toLocation(world, pos), activatingPlayers, searchCooldownTicks);
++
++        searchCooldownTicks = preSpawnerSearchEvent.getBackoffTicks();
++
++        if (preSpawnerSearchEvent.isCancelled()) return;
++        // Paper end
++
+         if (this.isNearPlayer(world, pos)) {
```

Comment:
Is it intended that this still checks for nearbyPlayers instead of using activatingPlayers?

Reply:
Not sure. It wasn't what my PR was attempting to achieve but I guess this could be implemented instead of the previous? Thoughts?
```suggestion
         if (!activatingPlayers.isEmpty()) {
```

Reply:
This is fine, i don't except a getter to be mutable for a list (by default), and that wouldn't be useful here but for safety you can change the list to an unmodifiable one for the getter only.
The only missing thing is a small comment saying you copied the logic from hasNearbyAlivePlayerThatAffectsSpawning or you can keep your change to avoid two loops but in both case the list shouldn't be mutable by the plugin.

Reply:
Basically what you mean is to allow for the list to be able to add/remove players (basically allow spawners to be activated at any range?). 

In my opinion I don't think this is a desirable outcome as basically it will allow spawner's to functionally be always active which in my case would be opposite of the PR.

Although, I could add something like the following (not sure if this is what you meant!), although I don't think it would be great.
```java
private List<Player> activatingPlayers = new ArrayList<>();
private boolean activatingPlayersModified = false;

public Collection<Player> getActivatingPlayers() {
    return Collections.unmodifiableList(activatingPlayers);
}

public void addPlayer(Player player) {
    activatingPlayers.add(player);
    activatingPlayersModified = true;
}

public void setActivatingPlayers(Collection<Player> activatingPlayers){
  //whatever logic to be implemented
}

public boolean hasActivatingPlayersBeenModified() {
    return activatingPlayersModified;
}

Reply:
You don't really need addPlayer here and the boolean but the goal is to allow a and c otherwise how you would do?
For your point c you could have a #setResult([Event.Result](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/event/Event.Result.html)) / e.getResult() that would bypass the player check when not set to default for example.

Reply:
So basically I won't need the cancellable in this case if I have set/getResult is what I'm understanding?

Additionally I don't know why the sync added all the other files. I tried reverting but it just added more files. Is there anyway to just remove them from this PR?

Reply:
Yes, this looks good